gpu/test/matrix/ell_kernels.cpp

```diff
@@ -0,0 +1,194 @@
+   and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <core/matrix/ell.hpp>
+
+
+#include <random>
+
+
+#include <gtest/gtest.h>
+
+
+#include <core/base/exception.hpp>
+#include <core/base/executor.hpp>
+#include <core/matrix/dense.hpp>
+#include <core/test/utils.hpp>
+#include "core/base/exception_helpers.hpp"
+
+
+namespace {
+
+
+class Ell : public ::testing::Test {
+protected:
+    using Mtx = gko::matrix::Ell<>;
+    using Vec = gko::matrix::Dense<>;
+
+    Ell() : rand_engine(42) {}
+
+    void SetUp()
+    {
+        ASSERT_GT(gko::GpuExecutor::get_num_devices(), 0);
+        ref = gko::ReferenceExecutor::create();
+        gpu = gko::GpuExecutor::create(0, ref);
+    }
+
+    void TearDown()
+    {
+        if (gpu != nullptr) {
+            ASSERT_NO_THROW(gpu->synchronize());
+        }
+    }
+
+    std::unique_ptr<Vec> gen_mtx(int num_rows, int num_cols, int min_nnz_row)
+    {
+        return gko::test::generate_random_matrix<Vec>(
+            ref, num_rows, num_cols,
+            std::uniform_int_distribution<>(min_nnz_row, num_cols),
+            std::normal_distribution<>(-1.0, 1.0), rand_engine);
+    }
+
+    void set_up_apply_data()
+    {
+        mtx = Mtx::create(ref);
+        mtx->copy_from(gen_mtx(532, 231, 1));
+        expected = gen_mtx(532, 1, 1);
+        y = gen_mtx(231, 1, 1);
+        alpha = gko::initialize<Vec>({2.0}, ref);
+        beta = gko::initialize<Vec>({-1.0}, ref);
+        dmtx = Mtx::create(gpu);
+        dmtx->copy_from(mtx.get());
+        dresult = Vec::create(gpu);
+        dresult->copy_from(expected.get());
+        dy = Vec::create(gpu);
+        dy->copy_from(y.get());
+        dalpha = Vec::create(gpu);
+        dalpha->copy_from(alpha.get());
+        dbeta = Vec::create(gpu);
+        dbeta->copy_from(beta.get());
+    }
+
+
+    void set_up_apply_data_with_stride(int stride, int max_nonzeros_per_row)
```

Comment:
Do we need two functions here?

Isn't it enough to use default parameters like this:

```c++
void set_up_apply_data(int stride = 0, int max_nonzeros_per_row = 0)
```

and we get both behaviors with a single function. Or am I missing something?

Reply:
I have combined the two functions into one.